desc/io/equilibrium_io.py

```diff
@@ -71,7 +74,79 @@ def load(load_from, file_format=None):
     return obj
 
 
-class IOAble(ABC):
+def _unjittable(x):
+    # strings and functions can't be args to jitted functions, and ints/bools are pretty
+    # much always flags or array sizes which also need to be a compile time constant
+    if isinstance(x, (list, tuple)):
+        return any([_unjittable(y) for y in x])
+    if isinstance(x, dict):
+        return any([_unjittable(y) for y in x.values()])
+    if hasattr(x, "dtype"):
+        return np.issubdtype(x.dtype, np.integer) or np.issubdtype(x.dtype, np.bool_)
+    return isinstance(x, (str, types.FunctionType, bool, int, np.int_))
+
+
+def _make_hashable(x):
+    # turn unhashable ndarray of ints nto a hashable tuple
+    if hasattr(x, "shape"):
+        return ("ndarray", x.shape, tuple(x.flatten()))
+    return x
+
+
+def _unmake_hashable(x):
+    # turn tuple of ints and shape to ndarray
+    if isinstance(x, tuple) and x[0] == "ndarray":
+        return np.array(x[2]).reshape(x[1])
+    return x
+
```

Comment:
it could be better to make a hashable array class as outlined here https://github.com/google/jax/issues/4572#issuecomment-709809897

Reply:
I looked into something like that and it's probably overkill. The real issue is hashing the actual values of the array, and converting it to a tuple and hashing that seemed the fastest option of the things I tried.

Reply:
I was thinking just copying the [tuple hash](https://github.com/python/cpython/blob/6e9f83d9aee34192de5d0ef7285be23514911ccd/Objects/tupleobject.c#L319) in python, but it's fine if this works